### PR TITLE
fff: add module

### PR DESCRIPTION
**Pull request by Gigahawk**
### Checklist

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/rycee/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).
    - Test not added for `fff` directly, since all it's configuration is currently handled by `home.sessionVariables`.

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/rycee/home-manager/blob/master/CONTRIBUTING.md#commits) for more information and [recent commit messages](https://github.com/rycee/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/rycee/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

**Comment by Gigahawk** [comment]
Didn't add a test for the `fff` module itself, since right now it's just using `home.sessionVariables`.

**Comment by Gigahawk** [comment]
Not sure if `xsession` is the right spot to handle this config, since switching configurations don't apply until the user relogs, if there's a better solution please let me know.

File: modules/programs/fff.nix [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+      FFF_KEY_RENAME = cfg.keyBindings.rename;
+      FFF_KEY_MKDIR = cfg.keyBindings.mkdir;
+      FFF_KEY_MKFILE = cfg.keyBindings.mkfile;
+
+      FFF_KEY_ATTRIBUTES = cfg.keyBindings.attributes;
+
+      FFF_KEY_HIDDEN = cfg.keyBindings.hidden;
+    };
+
+    home.sessionFunctions = { "${cfg.cdOnExit.name}" = cfg.cdOnExit.commands; };
```
Comment (rycee) [inline modules/programs/fff.nix:576]: Replace by
```
programs.bash.initExtra = mkIf cfg.enableBashIntegration ''
  ${cfg.cdOnExit.name}() {
    ${cfg.cdOnExit.commands}
  }
'';
```

And similarly for Fish.

File: modules/programs/fff.nix [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+        '';
+      };
+      trashCmd = mkOption {
+        type = types.str;
+        default = "mv";
+        description = ''
+          Command to use when trashing a file
+        '';
+      };
+      favorite1 = mkOption {
```
Comment (rycee) [inline modules/programs/fff.nix:122]: Use list or attribute set instead of explicit numbers in option name.

File: modules/home-environment.nix [hunk 3]
```diff
@@ -198,6 +198,24 @@ in
       '';
     };
 
+    home.sessionFunctions = mkOption {
```
Comment (rycee) [inline modules/home-environment.nix:201]: Remove these changes.

File: modules/lib/shell.nix [hunk 4]
```diff
@@ -4,8 +4,34 @@ rec {
   # Produces a Bourne shell like variable export statement.
   export = n: v: ''export ${n}="${toString v}"'';
 
+  # Produces a Bourne shell like function declaration.
```
Comment (rycee) [inline modules/lib/shell.nix:7]: Unnecessary changes, can be removed.

File: modules/programs/bash.nix [hunk 5]
```diff
@@ -6,9 +6,7 @@ let
 
   cfg = config.programs.bash;
 
-in
```
Comment (rycee) [inline modules/programs/bash.nix:9]: Remove auto formatting for `bash.nix`, it should not be auto formatted.

File: modules/programs/fff.nix [hunk 6]
```diff
@@ -0,0 +1,582 @@
...
+
+let
+
+  cfg = config.programs.fff;
+
+in {
+  meta.maintainers = [ maintainers.gigahawk ];
+
+  options.programs.fff = {
+    enable = mkEnableOption "A simple file manager written in bash";
```
Comment (rycee) [inline modules/programs/fff.nix:13]: Make sure the description looks ok in manual.

File: modules/programs/fff.nix [hunk 7]
```diff
@@ -0,0 +1,582 @@
...
+        default = ''
+          fff $argv
+          cd (cat ${cfg.cdOnExit.helperFile})
+        '';
+        description = ''
+          Fish commands used to start fff with cdOnExit enabled.
+        '';
+      };
+
+      helperFile = mkOption {
```
Comment (rycee) [inline modules/programs/fff.nix:56]: Seems like an unnecessary option, it can be removed.

File: modules/programs/fff.nix [hunk 8]
```diff
@@ -0,0 +1,582 @@
...
+        type = types.str;
+        default = "${config.home.homeDirectory}/.cache/fff/.fff_d";
```
Comment (rycee) [inline modules/programs/fff.nix:58]:
```suggestion
        default = "${config.xdg.cacheHome}/fff/.fff_d";
```

File: modules/programs/fff.nix [hunk 9]
```diff
@@ -0,0 +1,582 @@
...
+        description = ''
+          Location of CD on exit helper file
+        '';
+      };
+    };
+
+    config = {
```
Comment (rycee) [inline modules/programs/fff.nix:65]: Avoid the use of config in option names, it is already a very overloaded term.

Comment (rycee) [inline modules/programs/fff.nix:65]: I think you can remove most of the options here (or possibly choose a few of the most important options). Then introduce a `settings` option of type `attrsOf str` as per [RFC 42](https://github.com/Infinisil/rfcs/blob/config-option/rfcs/0042-config-option.md) that users can use.

File: modules/programs/zsh.nix [hunk 10]
```diff
@@ -6,17 +6,18 @@ let
```
Comment (rycee) [inline modules/programs/zsh.nix:6]: This file should not be auto formatted, can restore it as it were.

**Comment by rycee** [comment]
Thanks for the contribution! I've added a few comments.

I don't quite understand what you mean by xsession, there is nothing related to X in the PR from what I can see. If you mean the session variables in general then yes, that is unfortunate but necessary. The session variables are tied to the login session and cannot be reliably changed without logging out and back in.

File: modules/programs/fff.nix (hunk 2 above)
Comment (Gigahawk) [reply modules/programs/fff.nix]: User's could conceivably want to set favorites on a specific number without setting ones prior, so an attribute set was used.

File: modules/programs/fff.nix (hunk 7 above)
Comment (Gigahawk) [reply modules/programs/fff.nix:56]: I don't imagine a lot of people needing to change this from the default, but I also don't know if I would deem it unnecessary.
I figure someone might want to configure this if it's listed as an option upstream.

File: modules/programs/fff.nix (hunk 9 above)
Comment (Gigahawk) [reply modules/programs/fff.nix]: I'm not quite sure how exactly to apply that RFC to this situation, configs are meant to be exported as shell variables, it wouldn't make a lot of sense to have these settings exported to a separate file and then somehow sourced as part of shell start up.

The motivation for having the options explicitly defined is to prevent arbitrary settings from being set here since they will end up as global shell variables.

File: modules/programs/fff.nix (hunk 6 above)
Comment (Gigahawk) [reply modules/programs/fff.nix:13]: Sorry if this is a stupid question, but how do I build the manual?
I can't seem to find any documentation on how to build it.

File: modules/programs/fff.nix [hunk 11]
```diff
@@ -0,0 +1,594 @@
...
+        ${cfg.cdOnExit.commands}
+      }
+    '';
+
+  in mkIf cfg.enable {
+    home.packages = [ pkgs.fff ];
+
+    programs.bash.initExtra = mkIf cfg.cdOnExit.enable cdOnExitFunc;
+    programs.zsh.initExtra = mkIf cfg.cdOnExit.enable cdOnExitFunc;
+    programs.fish.interactiveShellInit = mkIf cfg.cdOnExit.enable cdOnExitFunc;
```
Comment (Gigahawk) [inline modules/programs/fff.nix:592]: Fix this

**Comment by Gigahawk** [comment]
Bump

File: modules/programs/fff.nix (hunk 6 above)
Comment (rycee) [reply modules/programs/fff.nix:13]: Either
```console
$ man home-configuration.nix
```
for the man page or add `manual.html.enable = true` to your configuration and run
```console
$ home-manager-help
```
for the HTML version.

File: modules/programs/fff.nix [hunk 12]
```diff
@@ -0,0 +1,594 @@
...
+        default = ".";
+        description = ''
+          Key to toggle hidden files.
+        '';
+      };
+    };
+  };
+
+  config = let
+    cdOnExitStr = if cfg.cdOnExit.enable then
```
Comment (rycee) [inline modules/programs/fff.nix:493]: Can remove this.

File: modules/programs/fff.nix [hunk 13]
```diff
@@ -0,0 +1,594 @@
...
+    cdOnExit = {
+      enable = mkOption {
+        type = types.bool;
+        default = true;
+        description = ''
+          Change directory to current folder on exit
+        '';
+      };
+
+      name = mkOption {
```
Comment (rycee) [inline modules/programs/fff.nix:24]: Maybe I'm missing something but I wasn't able to find any place where this option is used.

File: modules/programs/fff.nix (hunk 7 above)
Comment (rycee) [reply modules/programs/fff.nix:56]: In general it is best to only expose explicit options for those settings that are absolutely necessary for getting a functional setup. Any other options that the user would like to set should be settable through an `extraConfig` option. See [3.2.3. Add only valuable options](https://rycee.gitlab.io/home-manager/index.html#sec-guidelines-valuable-options).

In this case I imagine `extraConfig` having the type `attrsOf (oneOf [bool int str])` and for convenience it could automatically prepend the `FFF_` part and upper case the attribute names. So, this option could be set by a user through
```
extraConfig = {
  cd_file = "${config.home.homeDirectory}/custom/path";
}
```

File: modules/programs/fff.nix [hunk 14]
```diff
@@ -0,0 +1,594 @@
...
+          Name of function used to call fff.
+        '';
+      };
+
+      commands = mkOption {
+        type = types.str;
+        default = ''
+          # cd to exited folder for bash and zsh
+          fff "$@"
+          cd "$(cat "''${XDG_CACHE_HOME:=''${HOME}/.cache}/fff/.fff_d")"
```
Comment (rycee) [inline modules/programs/fff.nix:37]: I think this could be
```suggestion
          cd "$(cat "$FFF_CD_FILE")" 
```
instead? Similar with `fishCommands` below.

File: modules/programs/fff.nix (hunk 7 above)
Comment (rycee) [reply modules/programs/fff.nix:56]: Actually, having looked through the options now I think you could move to a mostly [RFC 42](https://github.com/NixOS/rfcs/pull/42) style configuration by just having
```
settings = mkOption {
  type = with types; attrsOf (oneOf [bool int str]);
  apply =
    let
      normalizeName = n: if hasPrefix "FFF_" n then n else "FFF_" + toUpper n;
    in mapAttrs' (n: v: { name = normalizeName n; value = v; });
  default = {};
  example = {
    trash_cmd = "trash";
  };
  description = ''
    … see https://github.com/dylanaraps/fff#customization for available options …
    … an option FFF_ABC can here optionally be written abc …
  '';
}

keyBindings = mkOption {
  type = with types; attrsOf str;
  # Rest much like settings above but using FFF_KEY_ instead of FFF_
  …
}
```

Then in the `config` section have something like
```
home.sessionVariables =
  {
    FFF_CD_FILE = "${config.xdg.cacheHome}/fff/.fff_d";
    # Any other options that must be set.
  }
  // cfg.keyBindings
  // cfg.settings;
```
